crates/motionfile/src/motion_interpolator.rs

```diff
@@ -57,21 +68,23 @@ impl<T: Debug + Interpolate<f32>> MotionInterpolator<T> {
                 time_since_start,
             } => {
                 let current_frame = &self.frames[current_frame_index];
-                if current_frame
+                match current_frame
```

Comment:
I'm looking at the motion file conditions for the first time. If I understood it correctly, you are checking the conditions before entering the next frame. Couldn't you model it in the way that the conditions are checked in each advancement step. This decoupling from the frames would allow to immediately abort or wait in the middle of frames which sounds more powerful to me. What was your reasoning? :thinking:

Reply:
Well this sort of depends on the `Condition`. For the new `FallenAbort`, this would be much better, since we can save some time again by restarting the motion earlier. However for other conditions like `Stabilized` which waits until the angular velocity is small it only makes sense to do this at distinct time points (otherwise no dynamic motions would be possible).

Reply:
Maybe we want to have two classes of conditions? Conditions that are evaluated in each cycle of a frame and conditions that are a special kind of frame which can wait for a condition. Try to challenge the design and see whether it can be made more general s.t. you have more clear responsibilities and abstractions. We could also move this to Ideas for Next Iterations.

Reply:
I will experiment a bit with this

Reply:
I've now implemented sth. like that and it works great. However the state transitions in the MotionInterpolator are very messy (and it doesn't even yet correct automatically for non-existent conditions). What do you think of how I implemented the second condition type and do you know a way to make the state transitions cleaner?

Reply:
The `advance_by()` function is already very long and complex. It also violates the [Single Responsibility Principle](https://en.wikipedia.org/wiki/Single-responsibility_principle). It does too many low-level things. One mitigation would be to split it into multiple functions that handle different parts of the function. That would also allow to reuse code and reduce code duplication.

Other than that the state transitions are probably needed if you want to support your aimed features (aka your requirements) where each condition may have different outcomes. That requires you to react to all outcomes. Maybe you could extract the timeout-functionality and put it into a more central place s.t. you can also check it at a more central place. Currently each condition can implement special functionality which is maybe not needed and would make your code simpler.

Reply:
I've split up the condition checking and state transitions now